[export] Ban tensor constants in tracing.

zhxchen17 opened the pull request:
Summary:
There're two patterns resulting in the similar constant lifting effect during tracing:
```
def forward(self, x):
    return x + torch.tensor(...)
```
and
```
class Model(torch.nn.Module):
     def __init__(self):
         super().__init__()
         self.a = torch.tensor(...)
     def forward(self, x):
         return x + self.a
```
In theory and practice, it will be too implicit to promote them as buffers in tracing, due to two reasons:

1. They don't have a formal FQN.
2. They likely result in baked-in values without being noticed by users.
3. Their values can be easily mutated without being noticed by users.

Therefore we want to make a policy decision to ban any form of tensor constants to show up in the traced graph. To proceed in the tracing, users either need to register these constants explicitly as buffers, or use constructors to properly initialize them. In some cases, users can get away without even wrapping scalars inside a torch.tensor() call.

Test Plan: buck test mode/opt caffe2/test:test_export -- -r test_export_tensor_constants

Differential Revision: D53436364

suo commented on torch/export/_trace.py:
```diff
@@ -84,6 +85,16 @@ def _ignore_backend_decomps():
+        ExportErrorType.TENSOR_CONSTANT,
+        "Leaf tensor constants must be owned by a torch.nn.Module. Possbile reasons: "
+        + "1. Calling torch.tensor() in forward functions. "
+        + "2. Some tensor is not registered as a buffer / parameter in the exported torch.nn.Module. "
```
Please include an explicit fix suggestion, e.g.
```
Please register this tensor constant as a non-persistent buffer, like:
    self.register_buffer(..., persistent=False)
```

suo commented on torch/export/_trace.py:
```diff
@@ -433,12 +444,21 @@ def _export_non_strict(
+    if len(list(gm.named_buffers())) > 0:
+        # torch.tensor() call.
+        # TODO(zhxchen17) better error message.
+        raise _tensor_constants_error(None)
```
I think we shouldn't leave good error messages as a todo; it's like impossible for a user to tell where a tensor constant came from without a reasonable stack trace. 

1. For inline constants: shouldn't we have a stack trace on the get_attr node?
2. For module attributes: can we find the owning module in the hierarchy?

zhxchen17 committed:
[export] Ban tensor constants in tracing. (#119226)

Summary:
X-link: https://github.com/pytorch/executorch/pull/1844


There're two patterns resulting in the similar constant lifting effect during tracing:
```
def forward(self, x):
    return x + torch.tensor(...)
```
and
```
class Model(torch.nn.Module):
     def __init__(self):
         super().__init__()
         self.a = torch.tensor(...)
     def forward(self, x):
         return x + self.a
```
In theory and practice, it will be too implicit to promote them as buffers in tracing, due to two reasons:

1. They don't have a formal FQN.
2. They likely result in baked-in values without being noticed by users.
3. Their values can be easily mutated without being noticed by users.

Therefore we want to make a policy decision to ban any form of tensor constants to show up in the traced graph. To proceed in the tracing, users either need to register these constants explicitly as buffers, or use constructors to properly initialize them. In some cases, users can get away without even wrapping scalars inside a torch.tensor() call.

Test Plan: buck test mode/opt caffe2/test:test_export -- -r test_export_tensor_constants

Differential Revision: D53436364

thiagocrepaldi commented:
@titaiwangms FYI

ezyang commented:
I'm not sure I understand the claim here. Creating a fresh copy of the constant so that it cannot be subsequently mutated is exactly what `aten.lift_fresh` is supposed to do. What am I supposed to do if I legitimately need to embed a constant tensor inside of my model?

suo commented:
They should register a non persistent buffer, which is exactly what this is for (on phone so don't have the PR but tensor constants were the stated purpose). Then .to will actually work

suo commented:
I do think we may need to think about how to roll this out more incrementally, or allow some sort of opt-in here. Even if this is a questionable pattern, I have found it in several places so maybe a warning for now is more appropriate?

ezyang commented:
But I don't need .to to work. If I have a bare tensor constant, that wasn't taking a device argument on construction, then definitionally it was defined on CPU.

Or is the problem that we're burning in the value of the device argument to torch.tensor at trace time? This does seem like a legitimate problem to me, but it seems like we should solve this problem directly. IIRC, what I told @suo in the past is whenever you have device baked into your IR representation, if you want to move device you have to go through the IR and scrub them all to be your new device. The same admonition applies for constants.

suo commented:
> But I don't need .to to work. If I have a bare tensor constant, that wasn't taking a device argument on construction, then definitionally it was defined on CPU.

Yeah, although this is somewhat complicated by the presence of `torch.set_default_device` and other ways of forwarding devices to tensor constructors. But you're right that maybe the user never intended `.to()` to cast every tensor, probably device mismatches notwithstanding.

> Or is the problem that we're burning in the value of the device argument to torch.tensor at trace time?

This particular thing is not directly related to device burn-in, although it does make unconditionally rewriting the device more complicated (maybe the user specifically *wanted* that constant tensor to stay!).

Anyway, I think that maybe this is too common to really ban in tracing. I thought maybe it would be nice to enforce some regularity here and mistakenly thought that mostly only test code does this tensor constant thing, but I found some more (semi-)legitimate use cases for tensor constants in fbgemm so it might be too disruptive.
